source/system/ArchiveManager.js

```diff
@@ -0,0 +1,301 @@
+                        })
+                        .catch(function __handleDehydratedReadError(err) {
+                            throw new VError(err, `Failed rehydrating item from storage with key: ${key}`);
+                        })
+            )))
+            .catch(function __handleRehydrateError(err) {
+                throw new VError(err, "Failed rehydrating sources");
+            });
+    }
+
+    /**
+     * Unlock a source
+     * @param {String} id The ID of the source to unlock
+     * @param {String} masterPassword The password to unlock the source
+     * @returns {Promise} A promise that resolves once the source is unlocked
+     */
+    unlock(id, masterPassword) {
+        let source;
+        return Promise
+            .resolve()
+            .then(() => {
+                const index = this.indexOfSource(id);
+                if (index < 0) {
+                    throw new VError("Source not found for ID");
+                }
+                source = this.sources[index];
+                if (source.status !== SourceStatus.LOCKED) {
+                    throw new VError(`Source state invalid: ${source.status}`);
+                }
+                source.status = SourceStatus.PENDING;
+                return Promise.all([
+                    createCredentials.fromSecureString(source.sourceCredentials, masterPassword),
+                    createCredentials.fromSecureString(source.archiveCredentials, masterPassword)
+                ]);
+            })
+            .then(([sourceCredentials, archiveCredentials] = []) => {
+                return credentialsToSource(sourceCredentials, archiveCredentials, /* initialise */ false)
+                    .then(sourceInfo => {
+                        this._replace(source.id, Object.assign(
+                            sourceInfo,
+                            {
+                                id: source.id,
+                                name: source.name,
+                                status: SourceStatus.UNLOCKED,
+                                type: sourceCredentials.type
+                            }
+                        ))
+                    })
+                    .catch(function __handleCredentialsMapError(err) {
+                        throw new VError(err, "Failed mapping credentials to a source");
+                    });
+            })
+            .catch(function(err) {
+                throw new VError(err, `Failed to unlock source with ID: ${id}`);
+            });
+    }
+
+    /**
+     * Replace a source by its ID
+     * @protected
+     * @param {String} id The ID of the source
+     * @param {UnlockedArchiveManagerSource|LockedArchiveManagerSource} source The source to replace it with
+     */
+    _replace(id, source) {
+        const index = this.indexOfSource(id);
+        if (index < 0) {
+            throw new VError(`Failed replacing source: Source not found`);
+        }
+        this.sources[index] = source;
```

Comment:
Not that we're going for speed here, but I don't necessarily thing that mapping would be better because you have to check every single item in a loop. Besides that, if we yank the array out from under the manager, what happens with async operations that are running simultaneously? There are so many slow promises that unlocking 2 items at once with array replacement could definitely break one..